### ISSUE TITLE
as the for bpr loss function, is the expression is right in the code?

### ISSUE THREAD

**Issue by jackyhawk** [quora/qmf#27]
As for the loss function and updating the latent vector(parameter) ,

i find there is different implement in the code and in the pater 

in code:
  Double BPREngine::loss(const Double scoreDifference) const {
  return **log(1.0 + exp(-scoreDifference));**
}

in paper:
 
i just paste the rule here:
http://photo27.hexun.com/p/2019/0201/632421/b_vip_11118BF2E45A1C8033DC6DA3576F7A2C.jpg 
[http://photo27.hexun.com/p/2019/0201/632421/b_vip_11118BF2E45A1C8033DC6DA3576F7A2C.jpg](url)

you can get the rule here  

**Θ Θ + α1 +e−ex^−uij x^uij · @@Θx^uij + λΘΘ**

**Comment by albietz** [comment]
Hi, the function `loss` just gives you the loss for a given triplet, which in this case is given by the negative log-likelihood according to the BPR model. What you're showing in the image corresponds to gradient updates, which you can find in the `BPREngine::update` function (see https://github.com/quora/qmf/blob/master/qmf/bpr/BPREngine.cpp#L178).

Hope this helps,
Alberto

**Comment by jackyhawk** [comment]
Thanks very much Alberto.

And I just made a mistake, what i want to say is the lossderitive in the training presss,which is 
https://github.com/quora/qmf/blob/master/qmf/bpr/BPREngine.cpp#L243 
https://github.com/quora/qmf/blob/master/qmf/bpr/BPREngine.cpp#L183 

just as you mentioned in the update function.

however the lossDerivative is not the same as the highlight part  in the picture:

http://photo27.hexun.com/p/2019/0201/632421/b_vip_11118BF2E45A1C8033DC6DA3576F7A2C.jpg

i wonder whether we should use this expression 

 1.0 / (1.0 + exp(scoreDifference));

or  this expression
  exp(-scoreDifference) / (1.0 + exp(-scoreDifference));

 to calculate the loss deritive?

**Comment by albietz** [comment]
Those two expressions are in fact equal :)

**Comment by jackyhawk** [comment]
Got it, Thanks very much Alberto.
